### PR TITLE
Unit-testing for the server API

### PR DESCRIPTION
**Pull request by vxgmichel**
This PR is meant to be a discussion about unit-testing pytango.

As a starting point, 3 tests have already been implemented: `test_empty_device`, `test_set_state` and `test_set_status`. Running pytest gives the following result (works with python 2 and python 3):

```console
.../pytango $ pytest -v
================================================ test session starts ================================================
platform linux2 -- Python 2.7.6, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- ~/.virtualenvs/tango9/bin/python
cachedir: .cache
rootdir: /home/vinmic/pytango, inifile: 
collected 3 items 

test/test_server.py::test_empty_device PASSED
test/test_server.py::test_set_state PASSED
test/test_server.py::test_set_status PASSED

============================================= 3 passed in 10.20 seconds =============================================
```

As you can see the tests are really slow. It turns out the first test lasts a few milliseconds whereas the following tests last 5 seconds. After investigation, I realized the device server takes exactly 5 seconds to start for the last two tests. Does anyone have an explanation for that?

### CODE REVIEW

**Comment by vxgmichel** [comment]
Related: [tango-cs/bugs #819](https://sourceforge.net/p/tango-cs/bugs/819/)

**Comment by vxgmichel** [comment]
I'm now using the `--boxed` option of `pytest-xdist` to run each test in its own subprocess. The 16 tests are running in 0.58 s:

```
(tango9-python3) ~/pytango $ python setup.py test --addopts -v
running pytest
running egg_info
writing dependency_links to pytango.egg-info/dependency_links.txt
writing requirements to pytango.egg-info/requires.txt
writing pytango.egg-info/PKG-INFO
writing top-level names to pytango.egg-info/top_level.txt
reading manifest file 'pytango.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
running build_ext
copying build/lib.linux-x86_64-3.4/tango/_tango.cpython-34m.so -> tango
================================================ test session starts ================================================
platform linux -- Python 3.4.3, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- ~/.virtualenvs/tango9-python3/bin/python
cachedir: .cache
rootdir: ~/pytango, inifile: pytest.ini
plugins: xdist-1.15.0
collected 16 items 

test/test_server.py::test_empty_device PASSED
test/test_server.py::test_set_state[MOVING] PASSED
test/test_server.py::test_set_state[ON] PASSED
test/test_server.py::test_set_state[FAULT] PASSED
test/test_server.py::test_set_state[DISABLE] PASSED
test/test_server.py::test_set_state[INSERT] PASSED
test/test_server.py::test_set_state[EXTRACT] PASSED
test/test_server.py::test_set_state[OFF] PASSED
test/test_server.py::test_set_state[UNKNOWN] PASSED
test/test_server.py::test_set_state[ALARM] PASSED
test/test_server.py::test_set_state[RUNNING] PASSED
test/test_server.py::test_set_state[STANDBY] PASSED
test/test_server.py::test_set_state[CLOSE] PASSED
test/test_server.py::test_set_state[INIT] PASSED
test/test_server.py::test_set_state[OPEN] PASSED
test/test_server.py::test_set_status PASSED

============================================= 16 passed in 0.58 seconds =============================================

**Comment by johanfforsberg** [comment]
Excellent stuff!

**Comment by johanfforsberg** [comment]
What is the reason to put the utils in the tango package but keep the tests outside?

I am not sure of the names "test_utils.py" and "test_context.py", they suggest that the files contain tests and might be picked up by test runners as such, right?

**Comment by vxgmichel** [comment]
@johanfforsberg Well, I used the test structure of asyncio as a reference. I think it makes sense to keep the actual tests out of the package, but I know it is something that is often debated. About the names, it should be fine since we restrict the pytest discovery to the `test/` directory, but I don't mind changing them if you have something nicer.

File: test/test_server.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+from tango import GreenMode
+from tango import DevState, AttrWriteType
+from tango.server import Device, DeviceMeta
+from tango.server import command, attribute, device_property
+from tango.test_utils import DeviceTestContext
+
+# Pytest fixtures
+from tango.test_utils import state, typed_values, server_green_mode
+state, typed_values, server_green_mode
```
Comment (vxgmichel) [inline test/test_server.py]: I didn't want Flake-8 to complain about unused imports, but there's probably a better way.

**Pull request merged**